### PR TITLE
Client Side Ghost Bot

### PR DESCRIPTION
**Pull request by PepperLola**
### Description
Serves to demonstrate client-host communication through the server for one user. The "ghost" bot is controlled by the user and has only behaviours that take input and generate signals. These signals are sent through the server to the host who does the physics simulation and who applies these signals to their own copy of that user's robot (the "host" bot in this PR). After applying the signals, the host responds with the transform data of each node of the robot. The client ghost bot receives these transforms and applies them to the corresponding nodes on its own instance. The ghost bot still lags behind the host bot, and I think it might be worth seeing if we can optimize the server-client communication.

There is no management of host vs. client and the host/ghost bots are hardcoded into the server test mode. Once the other multiplayer PRs that are in progress are merged, we can go through and get everything working together so that hosts and clients are designated properly when on different machines.

### Objectives
- [X] Client sends signal data and receives transform data to apply

### Testing Done
- Tested with Dozer, 2471, 6443

[JIRA Issue](https://jira.autodesk.com/browse/AARD-1499)

### CODE REVIEW

File: api/Api/Aether/Lobby/LobbyClient.cs [hunk 1]
```diff
@@ -45,8 +47,8 @@ private class Inner : IDisposable {
             private readonly LobbyClientHandler _handler;
             public LobbyClientHandler Handler => _handler;
 
-            private ReaderWriterLockSlim _transformDataLock;
-            private Dictionary<ulong, ServerTransforms> _transformData;
+            public ReaderWriterLockSlim _transformDataLock;
+            public Dictionary<ulong, ServerTransforms> _transformData;
```
Comment (BrandonPacewic) [inline api/Api/Aether/Lobby/LobbyClient.cs:51]: If we want to make these public we need to update the variable names.

File: engine/Assets/Scripts/Modes/ServerTestMode.cs
Comment (BrandonPacewic) [inline engine/Assets/Scripts/Modes/ServerTestMode.cs:1]: What's our plan for leaving this file in on master?

File: engine/Assets/Scripts/SimObjects/RobotSimObject.cs [hunk 2]
```diff
@@ -281,6 +304,8 @@ public class RobotSimObject : SimObject, IPhysicsOverridable, IGizmo {
     }
 
     public override void Destroy() {
+        Client.Dispose();
+        Client = null;
```
Comment (BrandonPacewic) [inline engine/Assets/Scripts/SimObjects/RobotSimObject.cs:308]: Do we need to check if the client is `null` here?

File: engine/Assets/Scripts/Modes/ServerTestMode.cs
Comment (KyroVibe) [reply engine/Assets/Scripts/Modes/ServerTestMode.cs:1]: @BrandonPacewic We'll wait for your branch with the new modes then work from there

Comment (BrandonPacewic) [reply engine/Assets/Scripts/Modes/ServerTestMode.cs:1]: 👍

**Review by BrandonPacewic — APPROVED** [review]

**Comment by BrandonPacewic** [comment]
Just need to fix the conflicts now that we merged #883. Other than that I'm looking forward to putting this all together.

**Comment by BrandonPacewic** [comment]
I pushed one small change, I realized I left in some `using`s that were not needed inside of `LobbyServer.cs`.

**Comment by PepperLola** [comment]
Something from dev introduced a lot of lag again. All of the communication still works, but the host bot only started moving ~17s after I pressed a key and the ghost bot started receiving position updates after about a minute.

File: api/Api/Aether/Lobby/LobbyClient.cs [hunk 3]
```diff
@@ -367,27 +366,26 @@ internal class LobbyClientHandler : IDisposable {
                     var startRead = DateTime.UtcNow;
                     while (bytesRead != msgSize && (DateTime.UtcNow - startRead).TotalMilliseconds < READ_TIMEOUT_MS) {
                         bytesRead += stream.Read(msgBuf, bytesRead, msgSize - bytesRead);
-                        Thread.Sleep(10);
```
Comment (PepperLola) [inline api/Api/Aether/Lobby/LobbyClient.cs:370]: Is reading from a network stream particularly expensive? I don't see why we needed a `Thread.Sleep` and removing it completely removed the lag.

Comment (BrandonPacewic) [reply api/Api/Aether/Lobby/LobbyClient.cs:370]: Well if removing it improves the lag then its the sleep that seems to be more expensive.

File: engine/Assets/Scripts/Modes/Aether/ServerTestMode.cs [hunk 4]
```diff
@@ -39,7 +62,37 @@ public class ServerTestMode : IMode {
...
     }
 
-    public void Update() {}
+    public void Update() {
+        var transformsList = new List<ServerTransforms>();
+        if (_ghost is null || _host is null)
+            return;
+        if (_ghost.Client is null || _host.Client is null)
+            return;
+
```
Comment (LucaHaverty) [inline engine/Assets/Scripts/Modes/Aether/ServerTestMode.cs:71]: There are quite a few nullable reference type warnings around here that should probably be addressed when we integrate multiplayer into match mode.

**Review by LucaHaverty — APPROVED** [review]
The ghost bot seems to work well and not lag too far behind! I found a bug to do with exiting to the main menu then going back into server test mode, but because the point of this pr is just to get the ghost bot working, you probably don't have to worry about that now.

Nice work, I can't wait for match multiplayer!

**Comment by KyroVibe** [comment]
Looks like Multiplayer is in the integration stages now

**Pull request merged**